// @thalalabs/surf@0.0.13/build/cjs/hooks/useSubmitTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSubmitTransaction = void 0;
const react_1 = require("react");
const wallet_adapter_react_1 = require("@aptos-labs/wallet-adapter-react");
const aptos_1 = require("aptos");
function isRawPayload(payload) {
    return payload.rawPayload !== undefined;
}
const useSubmitTransaction = () => {
    const { signAndSubmitTransaction } = (0, wallet_adapter_react_1.useWallet)();
    const [isIdle, setIsIdle] = (0, react_1.useState)(true);
    const [isLoading, setIsLoading] = (0, react_1.useState)(false);
    const [data, setResult] = (0, react_1.useState)();
    const [error, setError] = (0, react_1.useState)();
    const idRef = (0, react_1.useRef)(0);
    function reset() {
        setIsIdle(true);
        setIsLoading(false);
        setResult(undefined);
        setError(undefined);
        idRef.current = idRef.current + 1;
    }
    async function submitTransaction(payload, { nodeUrl, onSuccess, onFailed, max_gas_amount }) {
        if (signAndSubmitTransaction === undefined) {
            return;
        }
        const id = idRef.current;
        if (!isIdle) {
            throw new Error('Transaction is already in progress');
        }
        setIsLoading(true);
        setIsIdle(false);
        const request = isRawPayload(payload) ? payload.rawPayload : payload;
        try {
            const { hash } = await signAndSubmitTransaction({
                type: 'entry_function_payload',
                ...request,
                arguments: request.arguments.map((arg) => {
                    if (Array.isArray(arg)) {
                        return arg.map((item) => item.toString());
                    }
                    else if (typeof arg === 'object') {
                        throw new Error(`a value of struct type: ${arg} is not supported`);
                    }
                    else {
                        return arg.toString();
                    }
                }),
            }, { max_gas_amount });
            const client = new aptos_1.AptosClient(nodeUrl);
            const result = (await client.waitForTransactionWithResult(hash, {
                checkSuccess: true,
            }));
            if (id === idRef.current) {
                try {
                    onSuccess?.(result);
                }
                catch (e) {
                    console.error(e);
                }
                setResult(result);
            }
        }
        catch (e) {
            if (id === idRef.current) {
                const error = e instanceof Error
                    ? e
                    : e
                        ? new Error(String(e))
                        : new Error('unknown error');
                try {
                    onFailed?.(error);
                }
                catch (e2) {
                    console.error(e2);
                }
                setError(error);
            }
        }
        finally {
            if (id === idRef.current) {
                setIsLoading(false);
            }
        }
    }
    return { submitTransaction, reset, isLoading, isIdle, data, error };
};
exports.useSubmitTransaction = useSubmitTransaction;
